wire-ios-request-strategy/Sources/Request Strategies/MLS/FetchBackendMLSPublicKeysRequestStrategy.swift

```diff
@@ -0,0 +1,111 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+public final class FetchBackendMLSPublicKeysRequestStrategy: AbstractRequestStrategy {
+
+    // MARK: - Properties
+
+    // Slow Sync
+
+    private unowned var syncStatus: SyncProgress
+
+    private let syncPhase: SyncPhase = .fetchingBackendMLSPublicKeys
+
+    private var isSlowSyncing: Bool { syncStatus.currentSyncPhase == syncPhase }
+
+    private var slowSyncTask: Task<Void, Never>?
+
+    // Action
+
+    private let actionHandler: FetchBackendMLSPublicKeysActionHandler
+    private let actionSync: EntityActionSync
+
+    // MARK: - Life cycle
+
+    public init(
+        withManagedObjectContext managedObjectContext: NSManagedObjectContext,
+        applicationStatus: ApplicationStatus,
+        syncProgress: SyncProgress
+    ) {
+        actionHandler = FetchBackendMLSPublicKeysActionHandler(context: managedObjectContext)
+        actionSync = EntityActionSync(actionHandlers: [actionHandler])
+        self.syncStatus = syncProgress
+
+        super.init(
+            withManagedObjectContext: managedObjectContext,
+            applicationStatus: applicationStatus
+        )
+
+        configuration = [
+            .allowsRequestsWhileUnauthenticated,
+            .allowsRequestsWhileOnline,
+            .allowsRequestsDuringQuickSync,
+            .allowsRequestsDuringSlowSync,
+            .allowsRequestsWhileWaitingForWebsocket,
+            .allowsRequestsWhileInBackground
+        ]
+    }
+
+    deinit {
+        slowSyncTask?.cancel()
+    }
+
+    // MARK: - Request
+
+    public override func nextRequestIfAllowed(for apiVersion: APIVersion) -> ZMTransportRequest? {
+        guard isSlowSyncing, slowSyncTask == nil else {
+            return nil
+        }
+
+        slowSyncTask = Task { [weak self, syncStatus, syncPhase] in
+            guard let self, !Task.isCancelled else { return }
+
+            WireLogger.mls.info("slow sync start fetch backend MLS public keys!")
+
+            do {
+                // perform action notifies the registered action handler `FetchBackendMLSPublicKeysActionHandler`.
+                // the action stay pending until in the operation loop creates and executes the next request.
+                // Here the task waits for the result and then continues to report to syncStatus.
+
+                var action = FetchBackendMLSPublicKeysAction()
+                let backendPublicKeys = try await action.perform(in: managedObjectContext.notificationContext)
+                let hasValidKeys = backendPublicKeys.removal.hasValidKeys()
+                BackendInfo.isMLSEnabled = hasValidKeys && (BackendInfo.apiVersion ?? .v0) >= .v5
```

Comment:
wouldn't the action failed with notAvailable if we're under v5? also is v5 still a valid version for us to start MLS instead of v6?

Reply:
Yes, if we are under v5, request won't be send. I removed this check